### PR TITLE
crossdock: enable java

### PR DESCRIPTION
**Pull request by vivekrsharma**
Enable java in crossdock to ensure java doesn't fall behind and maintaining the regression tests on java repo. Java doesn't support tchannel protocol and skip filters ensure to filter those tests. Fixes https://github.com/yarpc/yarpc-go/issues/704

### CODE REVIEW

**Comment by abhinav** [comment]
Why is this PR WIP?

**Comment by vivekrsharma** [comment]
I wasn't sure if I need to do something more. I'm planning to wait until build passes and then remove WIP.

**Comment by abhinav** [comment]
Okay it seems like this won't be possible. Java doesn't implement TChannel so it can't be called with `transport=tchannel` which will happen when we run crosssdock for YARPC Go.

Obvious options:

- Keep Java as a server but not a client in the Go tests.
- Add support for pruning/filtering to crossdock. Some sort of "exclude any entry from the test matrix where transport=tchannel and server=java or client=java". 

Depending on whether we have bandwidth for crossdock work, the first may be preferable at this time.

**Comment by vivekrsharma** [comment]
Okay, added java only on the client side and waiting for the build to complete.

**Comment by breerly** [comment]
@abhinav @vivekrsharma I think this might actually be worth then investment improving Crossdock.

To be sure that Java is ran to the fullest for all the improvements we will make to Go seems worthy.

Vivek, let's walk through the Crossdock improvement together - might be you can invest here next week :)

**Comment by vivekrsharma** [comment]
Thanks @breerly! That sounds good to me 👍

**Comment by vivekrsharma** [comment]
This is ready for review with one nit - crossdock uses custom image but I'm working on getting it in latest image.

File: docker-compose.yml [hunk 1]
```diff
@@ -2,7 +2,7 @@ version: '2.1'
 
 services:
     crossdock:
-        image: crossdock/crossdock
+        image: crossdock/crossdock:filter-crossdock
```
Comment (breerly) [inline docker-compose.yml]: Nice! Obv lets get the Crossdock PR in first so that we can land this pinned to master 👍

File: docker-compose.yml [hunk 2]
```diff
@@ -41,19 +41,31 @@ services:
             - AXIS_CTXAVAILABLETRANSPORTS=http;tchannel
 
             - BEHAVIOR_RAW=client,server,transport
+            - BEHAVIOR_SKIP_RAW=client=java+transport=tchannel,server=java+transport=tchannel
```
Comment (breerly) [inline docker-compose.yml]: Interesting.

**Comment by breerly** [comment]
@vivekrsharma please update the PR description with a background/motivation.

File: docker-compose.yml (hunk 2 above)
Comment (vivekrsharma) [inline docker-compose.yml]: Yes, this is very very powerful now. We can select and pick what tests to run while still respecting the "full blown cross product of axis" behavior.

File: docker-compose.yml [hunk 3]
```diff
@@ -41,19 +41,31 @@ services:
...
             - BEHAVIOR_JSON=client,server,transport
+            - BEHAVIOR_SKIP_JSON=client=java+transport=tchannel,server=java+transport=tchannel
             - BEHAVIOR_THRIFT=client,server,transport
+            - BEHAVIOR_SKIP_THRIFT=client=java+transport=tchannel,server=java+transport=tchannel
```
Comment (abhinav) [inline docker-compose.yml]: Actually, doesn't using the prefix `BEHAVIOR_SKIP_*` mean that I can't have a behavior named `skip_foo`? What about `SKIP_*`?

Also, the example in crossdock/crossdock#83 uses `SKIP_*`. Is this PR out of date?

Comment (vivekrsharma) [inline docker-compose.yml]: Yes, I'll update this PR and unpin master after I land https://github.com/crossdock/crossdock/pull/83

**Pull request merged**